core/controllers/base.py

```diff
@@ -492,31 +480,34 @@ def render_json(self, values):
         Args:
             values: dict. The key-value pairs to encode in the JSON response.
         """
-        self.response.content_type = b'application/json; charset=utf-8'
-        self.response.headers[b'Content-Disposition'] = (
-            b'attachment; filename="oppia-attachment.txt"')
-        self.response.headers[b'Strict-Transport-Security'] = (
-            b'max-age=31536000; includeSubDomains')
-        self.response.headers[b'X-Content-Type-Options'] = b'nosniff'
-        self.response.headers[b'X-Xss-Protection'] = b'1; mode=block'
+        self.response.content_type = 'application/json; charset=utf-8'
+        self.response.headers['Content-Disposition'] = (
+            'attachment; filename="oppia-attachment.txt"')
+        self.response.headers['Strict-Transport-Security'] = (
+            'max-age=31536000; includeSubDomains')
+        self.response.headers['X-Content-Type-Options'] = 'nosniff'
+        self.response.headers['X-Xss-Protection'] = '1; mode=block'
 
         json_output = json.dumps(values, cls=utils.JSONEncoderForHTML)
-        self.response.write('%s%s' % (feconf.XSSI_PREFIX, json_output))
+        self.response.write(
+            b'%s%s' % (feconf.XSSI_PREFIX, json_output.encode('utf-8')))
 
     def render_downloadable_file(self, values, filename, content_type):
         """Prepares downloadable content to be sent to the client.
 
         Args:
-            values: dict. The key-value pairs to include in the response.
+            values: bytes. The data of the downloadable file.
             filename: str. The name of the file to be rendered.
             content_type: str. The type of file to be rendered.
         """
-        self.response.headers[b'Content-Type'] = python_utils.convert_to_bytes(
-            content_type)
-        self.response.headers[
-            b'Content-Disposition'] = python_utils.convert_to_bytes(
-                'attachment; filename=%s' % filename)
-        self.response.write(values)
+        self.response.headers['Content-Type'] = content_type
+        self.response.headers['Content-Disposition'] = (
+            python_utils.UNICODE('attachment; filename=%s' % filename))
```

Comment:
In general, when should we use python_utils.UNICODE and when should we not?

Reply:
Here it is actually needed. `python_utils.UNICODE` is just a mapping of `str` so it should be used anywhere we want to transform something into a string. Except for places where we need to convert between string and bytes, where `encode` and `decode` should be used. Alternatively to for encoding `convert_to_bytes` can be used.

Since we will be getting rid of `python_utils` after the migration, this will the just be `str`.